### DISCUSSION TITLE
how tot rain the existing model with new datasets which has old plus new classes.

### DISCUSSION THREAD

**Discussion by shahbazahmed108** [ultralytics/ultralytics#7447] [Q&A]
I am using yolov8 in cmd 

sorry for the bad phrasing, let me explain the issue, I have trained a model on 

nc: 13

names: ["ARG","WM","CW","FW-TM","CL","MW","KOCHIA","PL","PL-AS","PW","RTF","BROME","THISTLE"]

and I got a best.pt model that is running nicely. I want to train my existing model on more images that have classes as 

names: [“PW”,“MW”,”PL”,”UB”,”CL”,”WM”,”WHEAT”,”HENBIT”,”NIGHTSHADE”,”BT”]

as you see some of the classes in old and new datasets are the same and some are new. the index of the same classes across two different datasets doesn't match.  I have gone through a thread that says that I can not train the existing model that I got from my 13 classes dataset with a new dataset to add more classes (common across two datasets and new in the new dataset). is it still the same? 

if yes, then i assume I have to combine both datasets and update my .yaml file to have both classes together. since some of the classes are common, does my below .yaml file look good? 

nc=18
names: ['ARG', 'WM', 'CW', 'FW-TM', 'CL', 'MW', 'KOCHIA', 'PL', 'PL-AS', 'PW', 'RTF', 'BROME', 'THISTLE', 'UB', 'WHEAT', 'HENBIT', 'NIGHTSHADE', 'BT']

now do I have to go into the label.txt file and change the index of classes to match new .yaml file one by one? if I describe the process correctly, and this is how it should go, my question is, is there a code to change the index of classes so they match the new .yaml file so I don't have to go into the label of each image?

I only followed the tutorial to train my first model and it is getting a little complicated as I am trying to add data. please let me know if the issue makes sense and also if there is a simple way to solve my problem. my goal is to train the existing model with new datasets that have old plus new classes. Thank you

**Comment by pderrenger** [comment]
@shahbazahmed108 to train your existing YOLOv8 model with a new dataset that includes both old and new classes, you're correct that you need to merge the datasets and create a unified `.yaml` file with all the classes. Your `.yaml` file with `nc=18` and the combined list of class names looks good.

Regarding the class indices in your label files, you will need to update them to match the new class order in your `.yaml` file. There isn't a built-in function in YOLOv8 to automatically update these indices, but you can write a simple script to do this. Here's a Python snippet that can help you update the class indices:

```python
import os
import glob

# Define your class mapping here. The keys are the old class indices, and the values are the new ones.
class_mapping = {
    0: 0,  # ARG
    1: 1,  # WM
    # Add the rest of your mappings here
    # ...
}

# Path to your labels directory
labels_dir = 'path/to/your/labels/'

# Loop through label files and update class indices
for label_file in glob.glob(os.path.join(labels_dir, '*.txt')):
    with open(label_file, 'r') as file:
        lines = file.readlines()
    
    with open(label_file, 'w') as file:
        for line in lines:
            parts = line.strip().split()
            class_index = int(parts[0])
            if class_index in class_mapping:
                parts[0] = str(class_mapping[class_index])
            file.write(' '.join(parts) + '\n')
```

Make sure to fill in the `class_mapping` dictionary with the correct mappings from your old class indices to the new ones. Also, replace `'path/to/your/labels/'` with the actual path to your label files.

Once you have updated the class indices in your label files, you can proceed to train your model with the combined dataset. Use the updated `.yaml` file for training, and the model should learn to recognize both the old and new classes.

Remember to back up your data before running any scripts that modify it, to prevent data loss in case of errors. Good luck with your training! 🚀
